### PR TITLE
fix(ZNTA-1116): use transaction when reading config from DB

### PR DESCRIPTION
**Pull request by seanf**
This change is [Reviewable](https://reviewable.io/reviews/zanata/zanata-server/1174)

### CODE REVIEW

**Comment by seanf** [comment]
@zanata-jenkins retest this please

**Comment by davidmason** [comment]
-(Awaiting review) +(Under review) 

<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary>
> @zanata-jenkins retest this please
</details>


---

Reviewed 4 of 4 files at r1, 2 of 3 files at r3.
Review status: 6 of 7 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[zanata-war/src/main/java/org/zanata/util/ScopeHelper.java, line 35 \[r3\]](https://reviewable.io:443/reviews/zanata/zanata-server/1174#-KIgIqQYUZBAUv_xPMvd:-KIgIqQYUZBAUv_xPMve:933317786) ([raw file](https://github.com/zanata/zanata-server/blob/47d47756e093459ddb8eab5d87eb9a9c09326cd5/zanata-war/src/main/java/org/zanata/util/ScopeHelper.java#L35)):*
> ```Java
>  * @author Sean Flanigan <a href="mailto:sflaniga@redhat.com">sflaniga@redhat.com</a>
>  */
> public class ScopeHelper {
> ```

Docs should indicate the responsibilities of this class.

---

*[zanata-war/src/main/java/org/zanata/util/ScopeHelper.java, line 43 \[r3\]](https://reviewable.io:443/reviews/zanata/zanata-server/1174#-KIgJW4S6h84uYHcL0hZ:-KIgJW4S6h84uYHcL0h_:-253867839) ([raw file](https://github.com/zanata/zanata-server/blob/47d47756e093459ddb8eab5d87eb9a9c09326cd5/zanata-war/src/main/java/org/zanata/util/ScopeHelper.java#L43)):*
> ```Java
>      * https://issues.jboss.org/browse/JBEAP-2526?focusedCommentId=13144080
>      * @param r code to execute with Request scope active
>      * @throws Exception
> ```

Under what circumstances is an exception thrown?
Is a more specific exception possible?

---

*[zanata-war/src/main/java/org/zanata/util/ScopeHelper.java, line 45 \[r3\]](https://reviewable.io:443/reviews/zanata/zanata-server/1174#-KIgJwlzoMd39dVCiXTU:-KIgJwlzoMd39dVCiXTV:-544209327) ([raw file](https://github.com/zanata/zanata-server/blob/47d47756e093459ddb8eab5d87eb9a9c09326cd5/zanata-war/src/main/java/org/zanata/util/ScopeHelper.java#L45)):*
> ```Java
>      * @throws Exception
>      */
>     public static void withRequestScope(RunnableEx r) throws Exception {
> ```

`runnable` would be a clearer name for future diffs than `r`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-server/1174#-:-KIgGLRIp8JorQ0Eizpd:-1866920137)*

**Comment by davidmason** [comment]
Reviewed 1 of 3 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-server/1174)*

**Comment by seanf** [comment]
Review status: 6 of 7 files reviewed at latest revision, 3 unresolved discussions.

---

*[zanata-war/src/main/java/org/zanata/util/ScopeHelper.java, line 35 \[r3\]](https://reviewable.io:443/reviews/zanata/zanata-server/1174#-KIgIqQYUZBAUv_xPMvd:-KIk7OjrjI4D3AOS_tQY:-499146009) ([raw file](https://github.com/zanata/zanata-server/blob/47d47756e093459ddb8eab5d87eb9a9c09326cd5/zanata-war/src/main/java/org/zanata/util/ScopeHelper.java#L35)):*
<details><summary><i>Previously, davidmason (David Mason) wrote…</i></summary>
> Docs should indicate the responsibilities of this class.
</details>
Done.

---

*[zanata-war/src/main/java/org/zanata/util/ScopeHelper.java, line 43 \[r3\]](https://reviewable.io:443/reviews/zanata/zanata-server/1174#-KIgJW4S6h84uYHcL0hZ:-KIk5EeNpaUkhl8LPoY9:1305315632) ([raw file](https://github.com/zanata/zanata-server/blob/47d47756e093459ddb8eab5d87eb9a9c09326cd5/zanata-war/src/main/java/org/zanata/util/ScopeHelper.java#L43)):*
<details><summary><i>Previously, davidmason (David Mason) wrote…</i></summary>
> Under what circumstances is an exception thrown?
> Is a more specific exception possible?
</details>
Naturally I would use a specific exception if it were possible. RunnableEx is just like a Runnable but it throws Exception, which is handy for lambdas in Java. So it could throw anything.

---

*[zanata-war/src/main/java/org/zanata/util/ScopeHelper.java, line 45 \[r3\]](https://reviewable.io:443/reviews/zanata/zanata-server/1174#-KIgJwlzoMd39dVCiXTU:-KIk7PJNaZIQm_aKQc5c:-499146009) ([raw file](https://github.com/zanata/zanata-server/blob/47d47756e093459ddb8eab5d87eb9a9c09326cd5/zanata-war/src/main/java/org/zanata/util/ScopeHelper.java#L45)):*
<details><summary><i>Previously, davidmason (David Mason) wrote…</i></summary>
> `runnable` would be a clearer name for future diffs than `r`
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-server/1174)*

**Comment by seanf** [comment]
@davidmason Could you review my changes please? I had to make more substantial changes to fix the underlying problem with scopes. ScopeHelper isn't actually used now, but it's not a lot of code, so I think it's worth keeping around.

@carlosmunoz @aeng One of you might need to take this over. It's passing tests now, so there shouldn't be much left to do.

<details><summary><i>Previously, davidmason (David Mason) wrote…</i></summary>
> -(Awaiting review) +(Under review) 
</details>


---

Review status: 4 of 8 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-server/1174#-:-KIl6n_kdj_6TqzLxVKB:-167361363)*

**Comment by carlosmunoz** [comment]
Reviewed 4 of 4 files at r1, 3 of 4 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[zanata-war/src/main/java/org/zanata/notification/EmailQueueMessageReceiver.java, line 73 \[r5\]](https://reviewable.io:443/reviews/zanata/zanata-server/1174#-KIzHrB229rNz-sg9jBw:-KIzHrB229rNz-sg9jBx:-826930965) ([raw file](https://github.com/zanata/zanata-server/blob/7ded4a3d89418bf8713144ec0607179c6f041115/zanata-war/src/main/java/org/zanata/notification/EmailQueueMessageReceiver.java#L73)):*
> ```Java
>                 ImmutableMap.of(
>                         LanguageTeamPermissionChangedEvent.class.getCanonicalName(),
>                         LanguageTeamPermissionChangeJmsMessagePayloadHandler.class);
> ```

Perhaps for the future we can use a more 'CDI' way of doing this with decorators.

Basically, we decorate the handler with an annotation which indicates the message type which it handles and that's how we pick the right implementation. For now this does the trick.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-server/1174)*

**Comment by davidmason** [comment]
:lgtm:

<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary>
> @davidmason Could you review my changes please? I had to make more substantial changes to fix the underlying problem with scopes. ScopeHelper isn't actually used now, but it's not a lot of code, so I think it's worth keeping around.
> 
> @carlosmunoz @aeng One of you might need to take this over. It's passing tests now, so there shouldn't be much left to do.
</details>


---

Reviewed 3 of 4 files at r4, 1 of 1 files at r5.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-server/1174#-:-KIzTbsJxk4zt2cjG6QW:1417088721)*

**Comment by definite** [comment]
It says "Reviewed 3 of 4 files at r4, 1 of 1 files at r5."
It also says "code-review/reviewable" -- 8 files reviewed.

Which should I believe?

**Comment by carlosmunoz** [comment]
@definite The green tick for 'code-review'reviewable' indicates that the reviewer has marked the pull request as completely reviewed. The others are just comments on each step of the review.

**Comment by djansen-redhat** [comment]
What is the status of this request? I see "work in progress" but "On QA", which are somewhat mutually exclusive.

**Comment by davidmason** [comment]
Reviewed 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[zanata-war/src/main/java/org/zanata/jpa/NoNestingTransactionStrategy.java, line 56 \[r6\]](https://reviewable.io:443/reviews/zanata/zanata-server/1174#-KJK0W8HYMIZ6guFxeiZ:-KJK0W8HYMIZ6guFxei_:-549400474) ([raw file](https://github.com/zanata/zanata-server/blob/91afdfa4cac3d13c08fc68e494025c9e1368d1f0/zanata-war/src/main/java/org/zanata/jpa/NoNestingTransactionStrategy.java#L56)):*
> ```Java
>                 // "WELD-000401 Failure while notifying an observer of event ..."
>                 log.warn("Attempted nested transaction", e);
>                 throw e;
> ```

Is the Weld logging issue fixed for this now?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-server/1174)*

**Comment by seanf** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[zanata-war/src/main/java/org/zanata/jpa/NoNestingTransactionStrategy.java, line 56 \[r6\]](https://reviewable.io:443/reviews/zanata/zanata-server/1174#-KJK0W8HYMIZ6guFxeiZ:-KJK0xizRDHISZ6pHZJD:-1402444731) ([raw file](https://github.com/zanata/zanata-server/blob/91afdfa4cac3d13c08fc68e494025c9e1368d1f0/zanata-war/src/main/java/org/zanata/jpa/NoNestingTransactionStrategy.java#L56)):*
<details><summary><i>Previously, davidmason (David Mason) wrote…</i></summary>
> Is the Weld logging issue fixed for this now?
</details>
I'm not sure what you mean. The limited logging performed by Weld for bad event observers might be intentional. I haven't reported it as an issue.

According to the functional tests for this PR, we don't have any other event observers triggering the Weld error message.

But if it happens again, I want the information logged, even if Weld thinks it's unimportant.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-server/1174)*

**Comment by davidmason** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[zanata-war/src/main/java/org/zanata/jpa/NoNestingTransactionStrategy.java, line 56 \[r6\]](https://reviewable.io:443/reviews/zanata/zanata-server/1174#-KJK0W8HYMIZ6guFxeiZ:-KJK5K1dfE-Jz7X4qKw0:1169421975) ([raw file](https://github.com/zanata/zanata-server/blob/91afdfa4cac3d13c08fc68e494025c9e1368d1f0/zanata-war/src/main/java/org/zanata/jpa/NoNestingTransactionStrategy.java#L56)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary>
> I'm not sure what you mean. The limited logging performed by Weld for bad event observers might be intentional. I haven't reported it as an issue.
> 
> According to the functional tests for this PR, we don't have any other event observers triggering the Weld error message.
> 
> But if it happens again, I want the information logged, even if Weld thinks it's unimportant.
</details>
Sorry, I actually read this diff completely backwards, so I thought you had removed the extra logging rather than adding it. How embarrassing :confounded: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-server/1174)*
